flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MonitorSource.java

```diff
@@ -0,0 +1,200 @@
+
+  private final TableLoader tableLoader;
+  private final RateLimiterStrategy rateLimiterStrategy;
+  private final long maxReadBack;
+
+  /**
+   * Creates a {@link org.apache.flink.api.connector.source.Source} which monitors an Iceberg table
+   * for changes.
+   *
+   * @param tableLoader used for accessing the table
+   * @param rateLimiterStrategy limits the frequency the table is checked
+   * @param maxReadBack sets the number of snapshots read before stopping change collection
+   */
+  public MonitorSource(
+      TableLoader tableLoader, RateLimiterStrategy rateLimiterStrategy, long maxReadBack) {
+    Preconditions.checkNotNull(tableLoader, "Table loader should no be null");
+    Preconditions.checkNotNull(rateLimiterStrategy, "Rate limiter strategy should no be null");
+    Preconditions.checkArgument(maxReadBack > 0, "Need to read at least 1 snapshot to work");
+
+    this.tableLoader = tableLoader;
+    this.rateLimiterStrategy = rateLimiterStrategy;
+    this.maxReadBack = maxReadBack;
+  }
+
+  @Override
+  public Boundedness getBoundedness() {
+    return Boundedness.CONTINUOUS_UNBOUNDED;
+  }
+
+  @Override
+  public TypeInformation<TableChange> getProducedType() {
+    return TypeInformation.of(TableChange.class);
+  }
+
+  @Override
+  Iterator<TableChange> createIterator() {
+    return new SchedulerEventIterator(tableLoader, null, maxReadBack);
+  }
+
+  @Override
+  SimpleVersionedSerializer<Iterator<TableChange>> getIteratorSerializer() {
+    return new SchedulerEventIteratorSerializer(tableLoader, maxReadBack);
+  }
+
+  @Override
+  public SourceReader<TableChange, GlobalSplit<TableChange>> createReader(
+      SourceReaderContext readerContext) throws Exception {
+    RateLimiter rateLimiter = rateLimiterStrategy.createRateLimiter(1);
+    return new RateLimitedSourceReader<>(super.createReader(readerContext), rateLimiter);
+  }
+
+  /** The Iterator which returns the latest changes on an Iceberg table. */
+  @VisibleForTesting
+  static class SchedulerEventIterator implements Iterator<TableChange> {
+    private Long lastSnapshotId;
+    private final long maxReadBack;
+    private final Table table;
+
+    SchedulerEventIterator(TableLoader tableLoader, Long lastSnapshotId, long maxReadBack) {
+      this.lastSnapshotId = lastSnapshotId;
+      this.maxReadBack = maxReadBack;
+      tableLoader.open();
+      this.table = tableLoader.loadTable();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return true;
+    }
+
+    @Override
+    public TableChange next() {
+      try {
+        table.refresh();
+        Snapshot currentSnapshot = table.currentSnapshot();
+        Long current = currentSnapshot != null ? currentSnapshot.snapshotId() : null;
+        Long checking = current;
+        TableChange event = new TableChange(0, 0, 0L, 0L, 0);
+        long readBack = 0;
+        while (checking != null && !checking.equals(lastSnapshotId) && ++readBack <= maxReadBack) {
```

Comment:
maxReadBack seems problematic. let's see here is the snapshot history (backwards left to right: from recent snapshot to oldest)
S7, S6, S5, S4, S3, S2,S1

lastSnapshotId=S2 before the while loop. 
maxReadBack=2

after the while loop
lastSnapshotId=S5

S3 and S4 will be skipped in the next loop. 

You would want to scan forward from the last position. You can see how the regular source throttle the enumeration. https://github.com/apache/iceberg/blob/main/flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java#L89

Reply:
The reason behind the `maxReadBack` is to avoid reading thousands of snapshots, which could be especially costly when running the Maintenance on tables which are never compacted before.

In every other case the expectation is that the monitor only reads a few snapshot at a time, since it is continuously running.

If the monitor is down for so long, that we reach the threshold, that usually means that all of the Maintenance Tasks will be triggered anyway, so we don't need to calculate the exact values.

Also, if we don't add several snapshots to the resulting TableChange, that is perfectly fine. While the Maintenance Tasks are triggered a bit later, this doesn't effect correctness, as the Tasks themselves read the current state of the table, and the collected TableChange is only used as a trigger.

Reply:
> While the Maintenance Tasks are triggered a bit later, this doesn't effect correctness, as the Tasks themselves read the current state of the table, and the collected TableChange is only used as a trigger.

this is true if we assume the scheduler would read the current state of the table for scheduling decision. Let's use small files compaction with table partitioned by ingestion time as an example. New commits typically touch 1 or 2 partitions. there is no need to scan the whole table metadata (which could be large) to determine the small files. compaction task can just check the new files for compaction candidates and possibly check the all data files from the partitions that the delta change/commit touched. that could reduce the scope of checking from whole table to small number of partitions.

What I am trying to say is that it can be beneficial to have incremental check and compaction (from last compacted snapshot to a new snapshot).

There are certainly some tasks requiring whole table check (like orphan files cleanup).

Reply:
`maxReadBack` is another config that users need to know how to config. it may not make much difference to check from the last snapshot to the current snapshot, since the snapshots are loaded anyway as a list in `TableMetadata`.

```
  private synchronized void ensureSnapshotsLoaded() {
    if (!snapshotsLoaded) {
      List<Snapshot> loadedSnapshots = Lists.newArrayList(snapshotsSupplier.get());
      loadedSnapshots.removeIf(s -> s.sequenceNumber() > lastSequenceNumber);

      this.snapshots = ImmutableList.copyOf(loadedSnapshots);
      this.snapshotsById = indexAndValidateSnapshots(snapshots, lastSequenceNumber);
```

Reply:
The Monitor/TriggerManager is the same for all of the Maintenance Task, they should be kept generic. The compaction jobs itself could store the state of the previous runs if we decide it is worth the complexity (list of the candidate, but not yet compacted files, snapshots which are scanned / snapshots which are not yet scanned)

The savings could be significant, if this is an
- Append only table
- Only a few partitions are written

It is not optional, if any of the following is true:
- The goal is delete file removal - any V2 table
- The distribution of the partitions are even

Do you think that we can continue this discussion independently of the MonitorSource?

Reply:
> `maxReadBack` is another config that users need to know how to config. it may not make much difference to check from the last snapshot to the current snapshot, since the snapshots are loaded anyway as a list in `TableMetadata`.

We need to have the added data files/added data size etc. for the snapshots - this means we need to read the manifest files. This could become costly after some size.

I have faced a situation with a table which was never compacted, that it had ~2000 snapshots (my test table left for the weekend with 2 min checkpointing period - 60*24*3/2) and it took serious time to read all of the snapshots (more than 20 min) - In this case the parallel read might help if we want "correct" results, but as I have mentioned before, we don't really need exact data in this case.

So having 100 as a default, and allow for override if necessary would be better. Normal users don't have to bother with it, advanced users can play around with it, if they want - also I would be fine with not allowing to change it from 100 if that is the community decision.